src/easy3d.c

```diff
@@ -0,0 +1,203 @@
+static void Easy3D_EngineSetup(void);
+
+void Easy3D_LoadModelFromPath(const u8 heapID, const char *path, NNSG3dResMdl **outModel, NNSG3dResFileHeader **outResource)
+{
+    *outResource = ReadFileToHeap(heapID, path);
+    NNS_G3D_NULL_ASSERT(*outResource);
+
+    NNSG3dResTex *texture = NNS_G3dGetTex(*outResource);
+    if (texture != NULL && Easy3D_IsTextureUploadedToVRam(texture) == FALSE) {
+        DC_FlushRange(*outResource, (*outResource)->fileSize);
+        NNS_G3dResDefaultSetup(*outResource);
+    }
+
+    *outModel = NNS_G3dGetMdlByIdx(NNS_G3dGetMdlSet(*outResource), 0);
+}
+
+void Easy3D_LoadModelFromResource(NNSG3dResMdl **outModel, NNSG3dResFileHeader **resource)
+{
+    NNSG3dResTex *texture = NNS_G3dGetTex(*resource);
+
+    if (texture != NULL && Easy3D_IsTextureUploadedToVRam(texture) == FALSE) {
+        DC_FlushRange(*resource, (*resource)->fileSize);
+        NNS_G3dResDefaultSetup(*resource);
+    }
+
+    *outModel = NNS_G3dGetMdlByIdx(NNS_G3dGetMdlSet(*resource), 0);
+}
+
+void Easy3D_InitRenderObjFromPath(const u8 heapID, const char *path, NNSG3dRenderObj *obj, NNSG3dResMdl **outModel, NNSG3dResFileHeader **outResource)
+{
+    Easy3D_LoadModelFromPath(heapID, path, outModel, outResource);
+    NNS_G3dRenderObjInit(obj, *outModel);
+}
+
+void Easy3D_InitRenderObjFromResource(NNSG3dRenderObj *renderObj, NNSG3dResMdl **model, NNSG3dResFileHeader **resource)
+{
+    Easy3D_LoadModelFromResource(model, resource);
+    GF_ASSERT(model);
+    NNS_G3dRenderObjInit(renderObj, *model);
+}
+
+BOOL Easy3D_IsTextureUploadedToVRam(NNSG3dResTex *texture)
+{
+    return (texture->texInfo.flag & NNS_G3D_RESTEX_LOADED) 
+        || (texture->tex4x4Info.flag & NNS_G3D_RESTEX4x4_LOADED);
+}
+
+void Easy3D_DrawRenderObj(NNSG3dRenderObj *renderObj, const VecFx32 *pos, const MtxFx33 *rot, const VecFx32 *scale)
+{
+    NNS_G3dGlbSetBaseTrans(pos);
+    NNS_G3dGlbSetBaseRot(rot);
+    NNS_G3dGlbSetBaseScale(scale);
+    NNS_G3dGlbFlush();
+    NNS_G3dDraw(renderObj);
+}
+
+void Easy3D_DrawRenderObjSimple(NNSG3dRenderObj *renderObj, const VecFx32 *pos, const MtxFx33 *rot, const VecFx32 *scale)
+{
+    NNS_G3dGlbSetBaseTrans(pos);
+    NNS_G3dGlbSetBaseRot(rot);
+    NNS_G3dGlbSetBaseScale(scale);
+    NNS_G3dGlbFlush();
+    NNS_G3dDraw1Mat1Shp(renderObj->resMdl, 0, 0, TRUE);
+    NNS_G3dGeFlushBuffer();
+}
+
+static GenericPointerData *Unk_021BFB0C = NULL;
+
+void Easy3D_Init(const u8 heapID)
+{
+    Unk_021BFB0C = sub_02024220(heapID, 0, 2, 0, 4, Easy3D_EngineSetup);
+}
+
+static void Easy3D_EngineSetup(void)
+{
+    GXLayers_EngineAToggleLayers(GX_PLANEMASK_BG0, TRUE);
+
+    G2_SetBG0Priority(1);
+    G3X_SetShading(GX_SHADING_TOON);
+    G3X_AntiAlias(TRUE);
+    G3X_AlphaTest(FALSE, 0);
+    G3X_AlphaBlend(TRUE);
+    G3X_EdgeMarking(FALSE);
+    G3X_SetFog(FALSE, GX_FOGBLEND_COLOR_ALPHA, GX_FOGSLOPE_0x8000, 0);
+    G3X_SetClearColor(GX_RGB(0, 0, 0), 0, 0x7fff, 63, FALSE);
```

Comment:
**question:** Would it be useful to mark `0x7FFF` as the maximum value for a RGB15 palette?

Reply:
The 0x7FFF he refers to a depth value, not a color. Point still stands though, a constant would be appropriate. Not sure what to call it though. As far as I know there isn't any constant defined for this in the SDK. `G3X_MAX_DEPTH` maybe or `G3X_DEPTH_MAX`?

Reply:
I like `G3X_DEPTH_MAX`.